cpp/src/fil/infer.cu

```diff
@@ -285,31 +341,57 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES>
                                                void* shared_workspace,
                                                size_t data_row_size)
     : finalize_block(shared_workspace, num_classes_),
-      per_class_margin(
+      per_class(
         (vec<NITEMS, float>*)((char*)shared_workspace + data_row_size)) {
     for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
-      per_class_margin[c] = vec<NITEMS, float>();  // initialize to 0.0f
+      per_class[c] = vec<NITEMS, float>();  // initialize to 0.0f
     // __syncthreads() is called in infer_k
   }
 
   __device__ __forceinline__ void accumulate(
     vec<NITEMS, float> single_tree_prediction, int tree) {
     // since threads are assigned to consecutive classes, no need for atomics
-    per_class_margin[tree % num_classes] += single_tree_prediction;
+    per_class[tree % num_classes] += single_tree_prediction;
     // __syncthreads() is called in infer_k
   }
 
   __device__ __forceinline__ void finalize(float* out, int num_rows,
                                            int num_outputs) {
-    // reduce per-class candidate margins to one best class candidate
-    // per thread (for each of the NITEMS rows)
-    vec<NITEMS, best_margin_label> best({-1, -INFINITY});
-
-    for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
-      best = ArgMax()(best, to_vec(c, per_class_margin[c]));
-
-    __syncthreads();  // free up per_class_margin[]
-    write_best_class_in_block(best, blockDim.x, out, num_rows);
+    if (num_outputs == 1) {  // will output class
+      // reduce per-class candidate margins to one best class candidate
+      // per thread (for each of the NITEMS rows)
+      vec<NITEMS, best_margin_label> best({-1, -INFINITY});
+
+      for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
+        best = Vectorized<cub::ArgMax>()(best, to_vec(c, per_class[c]));
+
+      __syncthreads();  // free up per_class[]
+      write_best_class_in_block(best, blockDim.x, out, num_rows);
+    } else {  // will output softmax-ed margins
+      vec<NITEMS, float> partial_max(-INFINITY);
+      for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
+        partial_max = Vectorized<cub::Max>()(partial_max, per_class[c]);
+      auto max =
+        block_all_max(partial_max, blockDim.x, &per_class[num_classes]);
+      // softmax phase 1: compute sum of exponents
+      vec<NITEMS, float> partial_soe;
+      for (int c = threadIdx.x; c < num_classes; c += blockDim.x) {
+        auto exponent = shifted_exp(per_class[c], max);
+        per_class[c] = exponent;
+        partial_soe += exponent;
+      }
+      // sum of exponents
+      auto soe =
+        block_all_sum(partial_soe, blockDim.x, &per_class[num_classes]);
+      // softmax phase 2: normalization happens while outputting
+      for (int i = 0; i < NITEMS; ++i) {
```

Comment:
Here, too, it would be better to separate softmax computation from writing out the result.